### ISSUE TITLE
Multiple inheritance

### ISSUE THREAD

**Issue by bernatfortet** [styled-components/styled-components#773]
Is or would the following be possible?
```js
export const Card = styled(Rounded, Column)
```

otherwise I have to do the following
```js
export const Card = styled(RoundedColumn)
```
Which is alright, but add more code.

**Comment by philpl** [comment]
I'd suggest you to use mixins with the `css` helper in this case, since it's not really an inheritance problem, but a compositional problem.

Let's say you put the code for rounded elements in a `rounded` mixin:

```js
import { css } from 'styled-components'

const rounded = css`
  border-radius: 5px;
`
```

And the one for columns in a `column` mixin:

```js
import { css } from 'styled-components'

const column = css`
  display: block;
  flex-shrink: 0;
  flex-grow: 0;
  flex-basis: 25%;
  width: 25%;
`
```

Then you'd define your card like this:

```js
import styled from 'styled-components'

const Card = styled.div`
  ${rounded}
  ${column}
`
```

We're still missing a section on mixins / composition in the new documentation, but an explanation of the helper itself can be found here: https://sc-next-docs.philpl.com/docs/api#css

**Comment by arecvlohe** [comment]
This is really neat!

**Comment by pragatigarud** [comment]
Great!!

**Comment by zenVentzi** [comment]
Neat indeed :)

**Comment by zenVentzi** [comment]
Neat indeed :)

**Comment by joetidee** [comment]
A quick question on using mixins. Every time a new instance of `Card` is created, does it:
a) Copy the mixins into each individual instance (Gorilla-banana situation), or
b) the mixins behave like functional composition and are placed on the prototype of Card
?

**Comment by kitten** [comment]
@joetidee this would be a question that’d be more suited for spectrum or a new issue, but let me elaborate on it here :)

First of all I’ll change the two categories of answers up a little bit. As a CSS-in-JS library they don’t really suit either disciplines we span perfectly.

The mixins in styled-components or similar libraries are rather different than pure compile-to-CSS mixins (think Sass), since they exist during JS runtime.

So A and B would sound more like the following to me:

A) Is the mixin computed into a combined CSS template in styledComponents, or

B) Are separate CSS classes created for mixins?

In styled-components the answer is: A. Mixins are flattened into a StyledComponent’s style “template.”

This doesn’t have any effect on your everyday-use. Even in SSR this is negligible as long as your SSR results eventually are gzip’d.

B is not an option in the answers you’ve outlined since they’re just tagged templates, so arrays of strings and interpolations.

They’re still functionally composed in that, without side effects they’re merged into the rest of a StyledComponent’s template.

But they’re also technically preevaluated at runtime, which means they’re “copied” into another template.

This is to preserve specificity, which trumps everything else in our case. We expect the mixin to always apply at the case you interpolate it into, respectless of what comes after. If something after the mixin overrides its rules that needs to be respected. Hence we merge the mixin into a component’s rules. Which is also important when a mixin is used multiple times.

At runtime we make sure that the tagged template literal css function already preformats the rules to evaluate non-function interpolations, so this is as efficient as we can make it.

Tl;dr: in SC mixins are inlined into the rest of your rules during runtime. This leads to duplication only during SSR, which is fine with gzip.

**Comment by Vishwaas** [comment]
@kitten  can we not do just template strings? it seems to work fine too, whats the advantage of css\`\`
```
const column = `
  display: block;
  flex-shrink: 0;
  flex-grow: 0;
  flex-basis: 25%;
  width: 25%;
`

// while importing
import styled from 'styled-components'

const Card = styled.div`
  ${column}
`
```

**Comment by probablyup** [comment]
You don’t get vendor prefixing that way

**Comment by Vishwaas** [comment]
@probablyup  sorry not sure I follow. You mean I cannot use -moz-, -webkit- etc?

Anyway I think the answer was in the docs, my bad
https://www.styled-components.com/docs/api#css 

![image](https://user-images.githubusercontent.com/6441192/60374256-c20e5300-99c0-11e9-97f2-ea07d51b8579.png)

**Comment by uriklar** [comment]
@kitten I created a follow up question in Spectrum. Linking it here for future reference. 
https://spectrum.chat/styled-components/general/reusing-classes-when-applying-css-as-mixin~4beb952c-4c20-4527-b0d8-b4e7ba311007

**Comment by renaudtertrais** [comment]
I was looking fot the same feature:
```js
export const Card = styled(Rounded, Column)``
```

While waiting for a support, this workaround seems to work with `v4`:
```js
export const Card = styled(Rounded).attrs({ as: Column })``
```

**Comment by rilexus** [comment]
`const Padding = css`
  padding-left: ${({ theme: { styles } }) => styles.space[4]};
  padding-top: ${({ theme: { styles } }) => styles.space[4]};
  padding-bottom: ${({ theme: { styles } }) => styles.space[4]};
  padding-right: ${({ theme: { styles } }) => styles.space[4]};
`;

const Background = css`
  background-color: ${({ theme: { styles } }: ThemedProps) =>
    styles.colors.darkestGray};
`;
const Shadow = css`
  box-shadow: 0px 6px 15px rgb(0 0 0 / 70%);
`;

const BorderRadius = css`
  border-radius: 0.5rem;
`;

function mergeCSS(...css): StyledInterface {
  const res = {};
  Object.keys(styled).forEach((tagName) => {
    res[tagName] = (...c) => styled[tagName]`
      ${css}
      ${c}
    `;
  });
  //@ts-ignore
  return res;
}

// @ts-ignore
const StyledTile = mergeCSS(Padding, Background, Shadow, BorderRadius).div`
  a {
    color: ${themedMainFontColor};
    cursor: default;
    text-decoration: none;
  }
`;`

**Comment by matthew-dean** [comment]
I still feel like this is worth exploring / supporting. If you export a CSS block, yes it works, but it repeats the CSS in the injected style sheet for no reason. It's the same CSS. Ideally, by composing two different components, it would preserve / apply their class names.

IMO this should just work:
```js
export const Card = styled(Rounded, Column)
```

It's just communicating that I want to apply those two generated class names, and create a third, and all 3 should be applied to my component.

**Comment by Vol4n3** [comment]
+1 for this feature. 
Exported CSS block make a generated CSS too big.
My workaround to reducing the size of generated CSS is: 
```
export const RoundedDiv = styled.div`code for Rounded`;
export const ColumnRounded = styled(RoundedDiv)`code for Column`;
export const Card = styled(ColumnRounded)`code for card`;
```
But it makes my code more complex.
We need this feature for simplifying code and optimize generated CSS. 
The tag type of component can't be multiple extends but cases can.
But I suggest 
```
const cssWithProps = css`color:${(props)=>props.color}`; // it is possible to made this ??

const  Card = styled.div.extends(cssWithProps,otherCss,other...)``;
//or maybe
const Card = styled(Base).extends(cssWithProps,otherCss,other...)``;
```

**Comment by welschmoor** [comment]
I've just checked. Using css`` increases bundle size. So there's little point in using css``.
Can't you guys during compilation make css`` block be its own class that is then added to elements, so we don't have to deal with the growing bundle size? 
Styled-components is unusable for large projects because of that, CSS just keeps on growing.
